user-bag-service/src/main/kotlin/io/baum/userbagservice/users/kafka/UserRepository.kt

```diff
@@ -0,0 +1,90 @@
+package io.baum.userbagservice.users.kafka
+
+import arrow.core.Either
+import io.baum.userbagservice.users.kafka.metadata.MetaDataService
+import io.baum.userbagservice.users.tools.toKafkaRecord
+import io.baum.userbagservice.users.tools.toList
+import io.baum.userbagservice.users.web.client.ApiClientFactory
+import io.baum.userbagservice.users.web.model.RemoteAddress
+import org.apache.kafka.common.serialization.Serdes
+import org.apache.kafka.common.serialization.Serializer
+import org.apache.kafka.streams.StoreQueryParameters
+import org.apache.kafka.streams.StreamsConfig
+import org.apache.kafka.streams.state.QueryableStoreTypes
+import org.apache.kafka.streams.state.ReadOnlyKeyValueStore
+import org.springframework.boot.autoconfigure.kafka.KafkaProperties
+import org.springframework.kafka.config.StreamsBuilderFactoryBean
+import org.springframework.stereotype.Repository
+import org.apache.kafka.streams.KeyValue
+
+@Repository
+class UserRepository(
+        private val streamsBuilder: StreamsBuilderFactoryBean,
+        private val metaDataService: MetaDataService,
+        private val apiClient: ApiClientFactory,
+        kafkaProperties: KafkaProperties
+) {
+    private val applicationServer = kafkaProperties.streams.properties[StreamsConfig.APPLICATION_SERVER_CONFIG]?.split(":")
+            ?: throw Exception("Unable to find spring.kafka.properties.application.server property")
+    private val thisHost = applicationServer[0]
+    private val thisPort = applicationServer[1]
+
+    private val keySerializer = Serdes.String().serializer()
+    private val encryptedUserStore = ClusteredStateStore<String, UserRecord?>("encrypted-users-store", keySerializer)
+
+    fun getUserById(id: String) = encryptedUserStore.get(id)
+
+    private inner class ClusteredStateStore<K, V>(private val name: String, private val keySerializer: Serializer<K>) {
+        val store: ReadOnlyKeyValueStore<K, V> by lazy {
+            streamsBuilder.kafkaStreams.store(
+                    StoreQueryParameters.fromNameAndType(
+                            name,
+                            QueryableStoreTypes.keyValueStore<K, V>()
+                    ).enableStaleStores()
+            )
+        }
+
+        fun get(key: K): Either<RemoteAddress, V?> {
+            val (host, port, isLocal) = metaDataService.streamsMetadataForStoreAndKey(name, key, keySerializer)
+
+            return if(isLocal) {
+                Either.Right(store.get(key))
+            } else {
+                Either.Left(RemoteAddress(host, port))
+            }
+        }
+
+        fun getAll(): List<KeyValue<K, V>> {
+            return store.all().toList()
+        }
+    }
+
+    fun getAllUsersLocal(): List<UserRecord?> = encryptedUserStore.getAll().map { it.value }
+
+    fun getAllUsers(): List<UserRecord?> {
```

Comment:
should this return a nullable?

Reply:
The `store.getAll()` command I use to get all the data from kafka seems to insist on returning nullable objects

Reply:
This is your API though, you can define how it looks.

If in your head it makes no sense to return a List<null>, just get rid of them